File: bwt.py
import time
import random
from numba import cuda


@cuda.jit
def table(d_gpuTable, d_length, d_glength):
    index = cuda.grid(1)
    stride = cuda.gridsize(1)
    i = index
    for i in range(index,stride,d_length):
        if( i < index ):
            d_gpuTable[i] = i
        else:
            d_gpuTable[i] = -1

def rotation(t):
    sortedRotations = []
    for i in range(len(t)):
        rotation = t[i:] + t[:i] ## banana$ --> anana$b --> banana just moving this to the front
        sortedRotations.append(rotation)

    sortedRotations = sorted(sortedRotations)
    return sortedRotations

def bwt(t):
    '''computes the BWT(t), where t is a string'''
    bwtStr = ""

    sortedRotations = []
    sortedRotations = rotation(t)
    for i in sortedRotations:
        bwtStr += i[-1]
    return bwtStr

def ranksList(text):
    ranks_lst = []
    a = 0
    c = 0
    g = 0
    t = 0
    d = 0
    for i in range(len(text)):
        if text[i] == '$':
            d += 1
            ranks_lst.append((text[i], d))
        if text[i] == 'A':
            a += 1
            ranks_lst.append((text[i], a))
        if text[i] == 'C':
            c += 1
            ranks_lst.append((text[i], c))
        if text[i] == 'G':
            g += 1
            ranks_lst.append((text[i], g))
        if text[i] == 'T':
            t += 1
            ranks_lst.append((text[i], t))
            
    return ranks_lst
            

def helperList(t):
    temp = sorted(t)
    helperStr = ''.join(temp)
    helper_lst = []
    a = 0
    c = 0
    g = 0
    t = 0
    d = 0
    for i in range(len(helperStr)):
        if helperStr[i] == '$':
            d += 1
            helper_lst.append((helperStr[i], d))
        if helperStr[i] == 'A':
            a += 1
            helper_lst.append((helperStr[i], a))
        if helperStr[i] == 'C':
            c += 1
            helper_lst.append((helperStr[i], c))
        if helperStr[i] == 'G':
            g += 1
            helper_lst.append((helperStr[i], g))
        if helperStr[i] == 'T':
            t += 1
            helper_lst.append((helperStr[i], t))
            
    return helper_lst
    

def reverseBwt(t):
    '''computes the inverse of the BWT, where t is a string'''
    ranks_lst = ranksList(t)
    helper_lst = helperList(t)
    
    text = ""
    
    for i in range(len(t)):
        if i == 0:
            currSuff = helper_lst[i]
            index = ranks_lst.index(currSuff)
            text += helper_lst[index][0]
            currSuff = helper_lst[index]
        else:
            index = ranks_lst.index(currSuff)
            text += helper_lst[index][0]
            currSuff = helper_lst[index]
    return text

f = open("hw3_test3.txt", "r")         # open the file
text =(''.join(random.choice('CGTA') for _ in range(10000))+'$')##f.readline().strip()            # text is on the first line
print("Original Text = ",text)
start = time.time()
print("BWT of Text   = ",bwt(text))
end = time.time()
print("Reverse BWT   = ",reverseBwt(bwt(text)))
print("Time Elapsed during BWT sort + assembly = ",(end-start))
# print(reverseBwt(text))              # expected output TACATCACGT$

File: bwt.cpp
#include <iostream>
#include <string>
#include<iterator>
#include<algorithm>
#include <vector>
#include <fstream>
#include <chrono>

using namespace std::chrono;
std::vector<std::pair<char, int>> createRankList(std::string inputStr)
{
	std::cout << "\nstart create rank list\n";
    std::vector<std::pair<char, int>> rank_lst;
    int a = 0;
    int c = 0;
    int g = 0;
    int t = 0;
    int d = 0;

    for (auto& s : inputStr)
    {
        char letter;
        int pos;
        switch (s)
        {
        case '$':
            d += 1;
            letter = s;
            pos = d;
            break;

        case 'A':
            a += 1;
            letter = s;
            pos = a;
            break;

        case 'C':
            c += 1;
            letter = s;
            pos = c;

            break;

        case 'G':
            g += 1;
            letter = s;
            pos = g;

            break;

        case 'T':
            t += 1;
            letter = s;
            pos = t;

            break;
        default:
            std::cout << " no case was chosen\n";
        }
        std::pair <char, int> rankPair = std::make_pair(letter,pos);
        rank_lst.push_back(rankPair);
    }
    return rank_lst;
}


std::vector<std::pair<char, int>> createHelperList(std::string inputStr)
{
    std::cout << "\nstart createhelper\n";
    std::vector<std::pair<char, int> > helper_lst;
    std::sort(inputStr.begin(), inputStr.end());
    //std::reverse(inputStr.begin(), inputStr.end());
    int a = 0;
    int c = 0;
    int g = 0;
    int t = 0;
    int d = 0;

    for (auto& s : inputStr)
    {
        char letter;
        int pos;
        switch (s)
        {
        case '$':
            d += 1;
            letter = s;
            pos = d;
            break;

        case 'A':
            a += 1;
            letter = s;
            pos = a;
            break;

        case 'C':
            c += 1;
            letter = s;
            pos = c;

            break;

        case 'G':
            g += 1;
            letter = s;
            pos = g;

            break;

        case 'T':
            t += 1;
            letter = s;
            pos = t;

            break;
        default:
            std::cout << " no case was chosen\n";
        }
        std::pair <char, int> rankPair = std::make_pair(letter, pos);
        helper_lst.push_back(rankPair);
      
    }
   // helper_lst.push_front(helper_lst.back());
   // helper_lst.pop_back();

    return helper_lst;
}

void computeSuffixArray(std::string input_text)
{
    std::cout << "\nstart computeSuffix\n";
    std::vector< std::pair< std::string, int> > suffixArr; // list of tuple < string text_input, index>

    int index = 0;

    for (int i = 0; i < input_text.length(); i++)
    {
        std::cout << input_text << std::endl;
        std::pair <std::string, int> suffPair = std::make_pair(input_text, index);
        suffixArr.push_back(suffPair);
  
        std::rotate(input_text.rbegin(), input_text.rbegin() + 1, input_text.rend());
        index++;

    }

    for (auto& x : suffixArr)
    {
        std::cout << " (" << x.first << "[" << x.first[0] << "]" << "," << x.second << ")";
    }
    std::cout << "we did it bois before sort print " << std::endl;
    sort(suffixArr.begin(), suffixArr.end());
   

    for (auto& x : suffixArr)
    {
        std::cout << " (" << x.first << "[" << x.first[0] << "]" << "," << x.second << ")";
    }
    std::cout << "we did it bois after sort" << std::endl;

}

std::vector<std::string> rotation(std::string input_text) 
{
    std::cout << "\nstart rotation func\n";
    std::cout << "\n this is the original input text: " << input_text << std::endl;
    std::vector<std::string> rotations;
    std::string first;

    for (int i = 0; i < input_text.length(); i++) {
        input_text = input_text.substr(1, input_text.length() - 1) + input_text.front();
        rotations.push_back(input_text);
    }

     std::cout << "==========this is all of the rotations before sort===============" << std::endl;
 
    for (auto const& i : rotations) {
        std::cout << i << std::endl;
    }



    std::cout << "==========this is all of the rotations===============" << std::endl;
    sort(rotations.begin(), rotations.end());
 
    for (auto const& i : rotations) {
        std::cout << i << std::endl;
    }
    return rotations;
}


std::string bwt(std::string input_text) 
{
    std::cout << "\nStart bwt func\n";

    std::vector<std::string> sortedlist = rotation(input_text);
    std::vector<std::string> finalList;
    std::string first;

    std::vector<std::string>::iterator it;

    std::string finalString;
    for (auto& x : sortedlist)
    {

        finalString.push_back(x.back());

        //std::cout << "this is finalString: " << finalString << std::endl;
    }
    std::cout << "==========" << std::endl;
    sort(finalList.begin(), finalList.end());
    std::cout << "\nthis is all strings in finalList\n";
    for (auto const& i : finalList) {
        std::cout << i << std::endl;
    }
    return finalString;
}

std::string reverseBwt(std::string inputStr)
{
    std::vector<std::pair<char, int> > rank_lst = createRankList(inputStr);
    std::vector<std::pair<char, int> > helper_lst = createHelperList(inputStr);
    
    std::cout << "==========RANK LIST PRINT OUT BEGIN ++++++++==" << std::endl;
    for (auto& x : rank_lst)
    {
        std::cout << " (" << x.first << "," << x.second << ")";
    }

    std::cout << "==========HELPER LIST PRINT OUT BEGIN ++++++++==" << std::endl;
    for (auto& x : helper_lst)
    {
        std::cout << " (" << x.first << "," << x.second << ")";
    }

    std::cout << "=======lists generated. starting loop=========\n"; 

    std::cout << "this is the size of inputStr: " << inputStr.size();
    std::string originalStr;
    int index;
    for(int i = 0; i < inputStr.size(); i++)
    {
        std::pair<char, int> currPair;
        if(i == 0)
	{
            currPair = helper_lst[i];
	}
	else
	{
	    currPair = helper_lst[index];
	}

            std::cout << "this is currPair: <suffix: " << currPair.first << ", index: " << currPair.second << ">\n"; 
	    std::vector<std::pair<char,int>>::iterator it;
            it = std::find(rank_lst.begin(), rank_lst.end(), currPair);
            index = it - rank_lst.begin();
            std::cout << "this is index or the element number of the rank_lst array: " << index << std::endl;

	    originalStr += helper_lst[index].first;
	    std::cout << "this is ogString: " << originalStr << std::endl;
	
    }
    return originalStr;
}
 
int main()
{
    std::vector<std::string> textString;
    std::ifstream input_stream;

    const char* str = "hw3_test3.txt";

    input_stream.open(str, std::ios_base::in);

    if (input_stream.is_open() == true)

    std::cout << "File is opened." << std::endl;

    else

    std::cout << "File could not be open!" << std::endl;
    
    std::string line;
    while(getline(input_stream, line))
    {
	 textString.push_back(line);
    }
    std::cout << "This is the file output: \n";;
    for(int i = 0; i < textString.size(); i++ )
    {
        std::cout  << textString[i] << " ";
    }
    std::cout << "\n======================================";
    /*std::string inputStr = "ACTGGCT$TGCGGC";

    std::vector<std::pair<char, int> > rank_lst = createRankList(inputStr);
    std::vector<std::pair<char, int> > helper_lst = createHelperList(inputStr);

    for (auto& x : rank_lst)
    {
        std::cout << " (" << x.first << "," << x.second << ")";
    }
    std::cout << "we did it bois " << std::endl;


    for (auto& x : helper_lst)
    {
        std::cout << " (" << x.first << "," << x.second << ")";
    }
    std::cout << "we did it bois " << std::endl;

    computeSuffixArray("banana$");
    */

    auto start = high_resolution_clock::now();
    std::string completeStr;
    completeStr = bwt(textString.front());
    std::cout << "\nThis is the complete bwt return string: " << completeStr << std::endl;
    std::cout << "starting reverseBWT\n";
    reverseBwt(bwt(textString.front()));
	
    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(stop - start);

    std::cout << duration.count() << std::endl;
	
	
    return 0;
}
